node/src/neighborhood/gossip_acceptor.rs

```diff
@@ -68,6 +73,100 @@ trait GossipHandler: NamedType + Send /* Send because lazily-written tests requi
     ) -> GossipAcceptanceResult;
 }
 
+struct IpChangeHandler {
+    remove_stream_sub: Recipient<RemoveStreamMsg>,
+    stream_shutdown_sub: Recipient<StreamShutdownMsg>,
+}
+
+impl NamedType for IpChangeHandler {
+    fn type_name(&self) -> &'static str {
+        "IpChangeHandler"
+    }
+}
+
+impl GossipHandler for IpChangeHandler {
+    // An IpChange must contain a single AGR; if it contains a NodeAddr, then (
+    //   it must have at least one port;
+    //   it must advertise the same IP address it came from
+    // );
+    // it must already be in our database as a next-door neighbor;
+    // and it must have a different IP address than the record in the database.
+
+    fn qualifies(&self, database: &NeighborhoodDatabase, agrs: &[AccessibleGossipRecord], gossip_source: SocketAddr) -> Qualification {
+        if agrs.len() != 1 {
+            return Qualification::Unmatched;
+        }
+        let agr = &agrs[0];
+        let existing_node = if let Some (existing_node) = database.node_by_key(&agr.inner.public_key) {
+            existing_node
+        }
+        else {
+            return Qualification::Unmatched;
+        };
+        if let Some (incoming_node_addr) = &agr.node_addr_opt {
+            if incoming_node_addr.ports().is_empty() {
+                return Qualification::Malformed(format!("Debut from {} for {} contained NodeAddr with no ports",
+                    gossip_source, agr.inner.public_key))
+            }
+            if incoming_node_addr.ip_addr() != gossip_source.ip() {
+                return Qualification::Malformed(format!("Debut from {} for {} has NodeAddr ({}) that does not match its source",
+                    gossip_source, agr.inner.public_key, incoming_node_addr))
+            }
+            if let Some (existing_node_addr) = existing_node.node_addr_opt() {
+                if incoming_node_addr.ip_addr() == existing_node_addr.ip_addr() {
+                    return Qualification::Unmatched
+                }
+            }
+        }
+        Qualification::Matched
+    }
+
+    fn handle(
+        &self,
+        _cryptde: &dyn CryptDE,
+        database: &mut NeighborhoodDatabase,
+        mut agrs: Vec<AccessibleGossipRecord>,
+        gossip_source: SocketAddr,
+        _connection_progress_peers: &[IpAddr],
+        _cpm_recipient: &Recipient<ConnectionProgressMessage>
+    ) -> GossipAcceptanceResult {
+        let source_agr = agrs.remove(0); // empty Gossip shouldn't get here
+        let mut old_addrs: Vec<SocketAddr> = vec![];
+        let mut db_node = database.node_by_key_mut(&source_agr.inner.public_key)
+            .expect("Node disappeared");
+        if let Some (node_addr) = db_node.metadata.node_addr_opt.as_ref() {
```

Comment:
This might be too much but I quite like this idea; you inspired me for that once (the difference is in initializing the variable `old_addrs`):

```
        let source_agr = agrs.remove(0); // empty Gossip shouldn't get here
        let mut db_node = database.node_by_key_mut(&source_agr.inner.public_key)
            .expect("Node disappeared");
        let old_addrs = if let Some (node_addr) = db_node.metadata.node_addr_opt.as_ref() {
            let ports_count = node_addr.ports().len();
            let mut old_addrs = Vec::with_capacity(ports_count);
            for port in node_addr.ports() {
                old_addrs.push (SocketAddr::new (node_addr.ip_addr(), port))
            }
            NodeAddr::new(&gossip_source.ip(), &node_addr.ports());
            db_node.metadata.node_addr_opt = Some (NodeAddr::new (&gossip_source.ip(), &node_addr.ports()));
        };
```

And I would've kind of expected you to write the iterator with `.for_each()`:
```
     node_addr.ports().for_each(|port|
                old_addrs.push (SocketAddr::new (node_addr.ip_addr(), port))
     )
```
...but maybe your style fits better to the necessity in the imperative call of `.push()` inside.

Reply:
Okay, I changed it, and fixed a couple of hidden problems as well.